pkg/services/sqlstore/migrations/ualert/alert_rule.go

```diff
@@ -0,0 +1,167 @@
+package ualert
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/grafana/grafana/pkg/util"
+)
+
+type alertRule struct {
+	OrgId           int64
+	Title           string
+	Condition       string
+	Data            []alertQuery
+	IntervalSeconds int64
+	Uid             string
+	NamespaceUid    string
+	RuleGroup       string
+	NoDataState     string
+	ExecErrState    string
+	For             duration
+	Updated         time.Time
+	Annotations     map[string]string
+	// Labels map[string]string (Labels are not Created in the migration)
+}
+
+func (m *migration) makeAlertRule(cond condition, da dashAlert, folderUID string) (*alertRule, error) {
+	migAnnotation := fmt.Sprintf(`{"dashboardUid": "%v", "panelId": %v, "alertId": %v}`, da.DashboardUID, da.PanelId, da.Id)
+
+	annotations := da.ParsedSettings.AlertRuleTags
+	if annotations == nil {
+		annotations = make(map[string]string, 1)
+	}
+	annotations["__migration__info__"] = migAnnotation
+
+	ar := &alertRule{
+		OrgId:           da.OrgId,
+		Title:           da.Name, // TODO: Make sure all names are unique, make new name on constraint insert error.
+		Condition:       cond.Condition,
+		Data:            cond.Data,
+		IntervalSeconds: ruleAdjustInterval(da.Frequency),
+		NamespaceUid:    folderUID, // Folder already created, comes from env var.
+		RuleGroup:       da.Name,
+		For:             duration(da.For),
+		Updated:         time.Now().UTC(),
+		Annotations:     annotations,
+	}
+	var err error
+	ar.Uid, err = m.generateAlertRuleUID(ar.OrgId)
+	if err != nil {
+		return nil, err
+	}
+
+	ar.NoDataState, err = transNoData(da.ParsedSettings.NoDataState)
+	if err != nil {
+		return nil, err
+	}
+
+	ar.ExecErrState, err = transExecErr(da.ParsedSettings.ExecutionErrorState)
+	if err != nil {
+		return nil, err
+	}
+
+	return ar, nil
+}
+
+func (m *migration) generateAlertRuleUID(orgId int64) (string, error) {
+	for i := 0; i < 20; i++ {
```

Comment:
I was thinking maybe best would be:
 - Assume there are no alert rules, or get all uids of existing alert rules
 - Since we know the number of alert_rules will equal the existing number of dashboard alerts, we know the number of UIDs
 - Since we know the number of UIDs we need (and can optionally fetch existing ones), we can pre-generate a unique list of UIDs to use in one shot - without hitting the DB each time.

Reply:
Made into issue https://github.com/grafana/alerting-squad/issues/121